src/Agent.Worker/TestResults/ResultsCommands.cs

```diff
@@ -0,0 +1,300 @@
+﻿using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    public class ResultsCommands : AgentService, ICommandExtension
+    {
+        private IExecutionContext _executionContext;
+        //publish test results inputs
+        private List<string> _testResultFiles;
+        private string _testRunner;
+        private bool _mergeResults;
+        private string _platform;
+        private string _configuration;
+        private string _runTitle;
+        private bool _publishRunLevelAttachments;
+
+        public void ProcessCommand(IExecutionContext context, Command command)
+        {
+            if (string.Equals(command.Event, WellKnownResultsCommand.PublishTestResults, StringComparison.OrdinalIgnoreCase))
+            {
+                try
+                {
+                    ArgUtil.NotNull(context, nameof(context));
+                    _executionContext = context;
+
+                    LoadPublishTestResultsInputs(command.Properties);
+
+                    // publish test results to VSO using the reader based on the test runner specified
+                    IResultReader resultsReader = GetTestResultReader(_testRunner);
+                    resultsReader.AddResultsFileToRunLevelAttachments = _publishRunLevelAttachments;
+                    PublishTestResults(resultsReader);
+                }
+                catch (Exception ex)
+                {
+                    //Do not fail the task.
+                    string message = ex.Message;
+                    if (ex.InnerException != null)
+                    {
+                        message += "\n";
```

Comment:
NewLine Char is different in different platform, use Environment.NewLine instead.